### PR TITLE
pkg/apis: Add validations for time settings of prometheus at CRD level

### PR DESCRIPTION
**Pull request by slashpai**
pkg/apis: Add validations for time settings of prometheus at CRD level

test: Update e2e test for CRD validation

This is needed to avoid invalid values from user which would
cause problem for application to load or not work as expected

Related-to https://github.com/prometheus-operator/prometheus-operator/issues/4524

Signed-off-by: Jayapriya Pai <slashpai9@gmail.com>

## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

**Note:**  I haven't full covered all time fields in this since alertmanager uses time.ParseDuration for a few and I don't want to mix different types in same PR. 

pkg/apis: Add validations for time settings of prometheus at CRD level

Regex validation is based on https://github.com/prometheus/common/blob/main/model/time.go#L186 included `0` also as valid

Regex test: https://regex101.com/r/2ciWaP/1

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [x] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Add validations for time settings of prometheus at CRD level
```

### CODE REVIEW

File: pkg/apis/monitoring/v1/types.go [hunk 1]
```diff
@@ -373,15 +375,21 @@ type PrometheusList struct {
...
+type Duration string
+
 // PrometheusSpec is a specification of the desired behavior of the Prometheus cluster. More info:
 // https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
 // +k8s:openapi-gen=true
 type PrometheusSpec struct {
 	CommonPrometheusFields `json:",inline"`
 	// Time duration Prometheus shall retain data for. Default is '24h' if
-	// retentionSize is not set, and must match the regular expression `[0-9]+(ms|s|m|h|d|w|y)`
+	// retentionSize is not set, and must match the regular expression `0|[0-9]+(ms|s|m|h|d|w|y)`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:389]: can you move this to another PR for better tracking?

File: pkg/apis/monitoring/v1/types.go [hunk 2]
```diff
@@ -1045,9 +1053,11 @@ type Endpoint struct {
 	// Optional HTTP URL parameters
 	Params map[string][]string `json:"params,omitempty"`
 	// Interval at which metrics should be scraped
-	Interval string `json:"interval,omitempty"`
+	// If not specified Prometheus' global scrape interval is used.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:1056]:
```suggestion
	// If not specified, the Prometheus global scrape interval is used.
```

File: pkg/apis/monitoring/v1/types.go [hunk 3]
```diff
@@ -1045,9 +1053,11 @@ type Endpoint struct {
...
+	Interval Duration `json:"interval,omitempty"`
 	// Timeout after which the scrape is ended
-	ScrapeTimeout string `json:"scrapeTimeout,omitempty"`
+	// If not specified Prometheus' global scrape timeout is used.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:1059]:
```suggestion
	// If not specified, the Prometheus global scrape timeout is used unless it is less than `Interval` in which the latter is used.
```

File: pkg/operator/validations.go [hunk 4]
```diff
@@ -46,10 +47,10 @@ func CompareScrapeTimeoutToScrapeInterval(scrapeTimeout, scrapeInterval string)
 		scrapeInterval = "30s"
 	}
 
-	if si, err = model.ParseDuration(scrapeInterval); err != nil {
+	if si, err = model.ParseDuration(string(scrapeInterval)); err != nil {
```
Comment (simonpasquier) [inline pkg/operator/validations.go:50]: same as for the size type PR, let's add a TODO comment for removing it in a future release.

File: pkg/apis/monitoring/v1/types.go [hunk 5]
```diff
@@ -373,15 +375,21 @@ type PrometheusList struct {
...
 // https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
 // +k8s:openapi-gen=true
 type PrometheusSpec struct {
 	CommonPrometheusFields `json:",inline"`
 	// Time duration Prometheus shall retain data for. Default is '24h' if
-	// retentionSize is not set, and must match the regular expression `[0-9]+(ms|s|m|h|d|w|y)`
+	// retentionSize is not set, and must match the regular expression `0|[0-9]+(ms|s|m|h|d|w|y)`
 	// (milliseconds seconds minutes hours days weeks years).
-	Retention string `json:"retention,omitempty"`
+	// +kubebuilder:default:="24h"
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:391]: ditto

Comment (slashpai) [reply pkg/apis/monitoring/v1/types.go:391]: I have added a todo on statefulset.go file to remove default

**Comment by slashpai** [comment]
I have updated with changes. I have covered only Prometheus time settings in this PR since some of Alertmanager fields uses time.Duration instead of model.Duration. Also will cover Thanos in separate PR

File: pkg/prometheus/promcfg.go [hunk 6]
```diff
@@ -470,14 +480,16 @@ func (cg *ConfigGenerator) Generate(
 
 	cfg := yaml.MapSlice{}
 
+	// TODO(slashpai): Remove this default assignment after v0.57 since this is set at CRD level
 	scrapeInterval := "30s"
 	if p.Spec.ScrapeInterval != "" {
-		scrapeInterval = p.Spec.ScrapeInterval
+		scrapeInterval = string(p.Spec.ScrapeInterval)
 	}
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:487]: this can be removed now since the returned object will have a default value of 30s already.

File: pkg/prometheus/promcfg.go [hunk 7]
```diff
@@ -470,14 +480,16 @@ func (cg *ConfigGenerator) Generate(
...
 
+	// TODO(slashpai): Remove this default assignment after v0.57 since this is set at CRD level
 	evaluationInterval := "30s"
 	if p.Spec.EvaluationInterval != "" {
-		evaluationInterval = p.Spec.EvaluationInterval
+		evaluationInterval = string(p.Spec.EvaluationInterval)
 	}
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:493]: same thing here

File: pkg/prometheus/statefulset.go [hunk 8]
```diff
@@ -40,6 +40,7 @@ import (
 	"github.com/prometheus-operator/prometheus-operator/pkg/operator"
 )
 
+// TODO(slashpai): Remove defaultRetention after v0.57 since default is set at CRD level
```
Comment (simonpasquier) [inline pkg/prometheus/statefulset.go:43]: could you move the comment next to where the const is used?

File: pkg/apis/monitoring/v1/types.go [hunk 9]
```diff
@@ -381,7 +388,8 @@ type PrometheusSpec struct {
 	// Time duration Prometheus shall retain data for. Default is '24h' if
 	// retentionSize is not set, and must match the regular expression `[0-9]+(ms|s|m|h|d|w|y)`
 	// (milliseconds seconds minutes hours days weeks years).
-	Retention string `json:"retention,omitempty"`
+	// +kubebuilder:default:="24h"
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:391]: I would add a note that time-based retention can be disabled by setting the field to "0" (but it requires a non-empty size-base retention to be effective).

File: pkg/prometheus/promcfg.go (hunk 6 above)
Comment (slashpai) [reply pkg/prometheus/promcfg.go]: We still have unit tests for this so I have added comment for this to remove after v0.57

File: pkg/prometheus/promcfg.go (hunk 7 above)
Comment (slashpai) [reply pkg/prometheus/promcfg.go]: We still have unit tests for this so I have added comment for this to remove after v0.57

File: pkg/apis/monitoring/v1/types.go [hunk 10]
```diff
@@ -381,7 +388,10 @@ type PrometheusSpec struct {
 	// Time duration Prometheus shall retain data for. Default is '24h' if
 	// retentionSize is not set, and must match the regular expression `[0-9]+(ms|s|m|h|d|w|y)`
 	// (milliseconds seconds minutes hours days weeks years).
-	Retention string `json:"retention,omitempty"`
+	// Time based retention can be disabled by setting Retention to `0` but it requires
+	// a non-empty size-based retention to be effective
+	// +kubebuilder:default:="24h"
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:393]: I realize that we can't handle the default value at this level: if `retentionSize` is not empty, the operator shouldn't set `retention` to 24h...

File: pkg/prometheus/promcfg.go [hunk 11]
```diff
@@ -377,71 +377,81 @@ func validateConfigInputs(p *v1.Prometheus) error {
 		}
 	}
 
+	// TODO(slashpai): Remove this validation after v0.57 since this is handled at CRD level
 	if p.Spec.Retention != "" {
```
[the remark this answers is not in the document]
Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:381]: Correcting myself: I think that we don't want a default value at the API level for `spec.retention`.

File: pkg/prometheus/promcfg.go (hunk 6 above)
Comment (simonpasquier) [reply pkg/prometheus/promcfg.go]: hmm I wish there was a way for the tests to get a Prometheus object with all defaults...

**Comment by slashpai** [comment]
I have updated with changes

**Comment by slashpai** [comment]
@simonpasquier Can you review?

**Review by simonpasquier — APPROVED** [review]
lgtm
